executor/src/subscriber.rs

```diff
@@ -120,144 +156,313 @@ impl Subscriber {
+            return payload;
+        }
+        let _timer = self.metrics.subscriber_remote_fetch_latency.start_timer();
+        let mut stagger = Duration::from_secs(0);
+        let mut futures = vec![];
+        for worker in workers {
+            let future = self.fetch_from_worker(stagger, worker, digest);
+            futures.push(future.boxed());
+            stagger += Duration::from_secs(1);
+        }
+        let (batch, _, _) = futures::future::select_all(futures).await;
+        batch
+    }
+
+    #[instrument(level = "debug", skip_all, fields(digest = % digest, worker_id = % worker_id))]
+    async fn try_fetch_locally(&self, digest: BatchDigest, worker_id: WorkerId) -> Option<Batch> {
+        let _timer = self.metrics.subscriber_local_fetch_latency.start_timer();
+        let worker = self.network.my_worker(&worker_id);
+        let payload = self.network.request_batch(digest, &worker).await;
+        match payload {
+            Ok(Some(batch)) => {
+                debug!("Payload {} found locally", digest);
+                self.metrics.subscriber_local_hit.inc();
+                return Some(batch);
             }
-            Err(err) => {
-                panic!("Irrecoverable error occurred while retrieving block payload: {err}");
+            Ok(None) => debug!("Payload {} not found locally", digest),
+            Err(err) => error!("Error communicating with out own worker: {}", err),
+        }
+        None
+    }
+
+    /// This future performs fetch from given worker
+    /// This future performs infinite retries with exponential backoff
+    /// You can specify stagger_delay before request is issued
+    #[instrument(level = "debug", skip_all, fields(stagger_delay = ? stagger_delay, worker = % worker, digest = % digest))]
+    async fn fetch_from_worker(
+        &self,
+        stagger_delay: Duration,
+        worker: NetworkPublicKey,
+        digest: BatchDigest,
+    ) -> Batch {
+        tokio::time::sleep(stagger_delay).await;
+        let max_timeout = Duration::from_secs(60);
+        let mut timeout = Duration::from_secs(10);
+        loop {
+            let deadline = Instant::now() + timeout;
+            let request_batch_guard =
+                PendingGuard::make_inc(&self.metrics.pending_remote_request_batch);
+            let payload =
+                tokio::time::timeout_at(deadline, self.safe_request_batch(digest, &worker)).await;
+            drop(request_batch_guard);
+            match payload {
+                Ok(Ok(Some(payload))) => return payload,
+                Ok(Ok(None)) => error!("[Protocol violation] Payload {} was not found at worker {} while authority signed certificate", digest, worker),
+                Ok(Err(err)) => debug!(
+                    "Error retrieving payload {} from {}: {}",
+                    digest, worker, err
+                ),
+                Err(_elapsed) => debug!("Timeout retrieving payload {} from {}",
```

Comment:
If this happens it seems we have a serious problem, don't we? Should we log a warn / error?

Reply:
Well I guess not necessarily - workers of other validators might not respond in time (for whatever reasons)

Reply:
I don't have strong opinion on log level here - do you know how we typically log timeouts in other components?